### PR TITLE
[SPARK-25636][CORE] spark-submit cuts off the failure reason when there is an error connecting to master

### PR DESCRIPTION
**Pull request by devaraj-kavali**
## What changes were proposed in this pull request?
Cause of the error is wrapped with SparkException, now finding the cause from the wrapped exception and throwing the cause instead of the wrapped exception.

## How was this patch tested?
Verified it manually by checking the cause of the error, it gives the error as shown below.

### Without the PR change

```
[apache-spark]$ ./bin/spark-submit --verbose --master spark://******
....
Error: Exception thrown in awaitResult:
Run with --help for usage help or --verbose for debug output

```

### With the PR change

```
[apache-spark]$ ./bin/spark-submit --verbose --master spark://******
....
Exception in thread "main" org.apache.spark.SparkException: Exception thrown in awaitResult:
        at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:226)
        at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:75)
        ....
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.io.IOException: Failed to connect to devaraj-pc1/10.3.66.65:7077
        at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:245)
       ....
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: devaraj-pc1/10.3.66.65:7077
        at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
        ....
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        ... 1 more
Caused by: java.net.ConnectException: Connection refused
        ... 11 more

```

### CODE REVIEW

**Comment by srowen** [comment]
The only thing I'm confused about is, right now when the `SparkException` is thrown instead of its cause, wouldn't you get a big stack trace starting with the `SparkException`? that `SparkException` is useful too. I don't see why it doesn't print in this case, do you? and can we fix that part.

**Comment by tgravescs** [comment]
Please update the title as it seems cut off.

**Comment by devaraj-kavali** [comment]
Thanks @srowen  for looking into this.

> ThreadUtils.scala

```
      case NonFatal(t) if !t.isInstanceOf[TimeoutException] =>
        throw new SparkException("Exception thrown in awaitResult: ", t)
```

Here the cause is getting wrapped as SparkException with the message. And in SparkSubmit.scala, it is just printing this message and discarding the caused exception.

> SparkSubmit.scala

```
      override def doSubmit(args: Array[String]): Unit = {
        try {
          super.doSubmit(args)
        } catch {
          case e: SparkUserAppException =>
            exitFn(e.exitCode)
          case e: SparkException =>
            printErrorAndExit(e.getMessage())
        }
```


The other option is to print the whole stack trace instead of just message here. Please let me know your thought, I can make change with this.

**Comment by srowen** [comment]
Yeah, that's the issue. Maybe change `printErrorAndExit` to `logError` the exception in this case. It's noisier to be sure but I think losing the stack trace here is significant. @vanzin I think you touched this part last; WDYT?

Otherwise this change doesn't address the fact that you'd still only print a little info about one cause, not all of them.

**Comment by vanzin** [comment]
Looking at the 2.3 code, it might be the right thing to just remove the handler for `SparkException` in `SparkSubmit.main`, and let it propagate.

```
    try {
      mainMethod.invoke(null, childArgs.toArray)
    } catch {
      case t: Throwable =>
        findCause(t) match {
          case SparkUserAppException(exitCode) =>
            System.exit(exitCode)

          case t: Throwable =>
            throw t
        }
    }
```

**Comment by vanzin** [comment]
ok to test

**Review by srowen — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 1]
```diff
@@ -74,20 +74,26 @@ trait TestPrematureExit {
...
     val thread = new Thread {
       override def run() = try {
         mainObject.main(input)
       } catch {
         // If exceptions occur after the "exit" has happened, fine to ignore them.
         // These represent code paths not reachable during normal execution.
-        case e: Exception => if (!exitedCleanly) throw e
+        case e: Exception =>
+          message = e.getMessage
+          if (!(exitedCleanly || message.contains(searchString))) {
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: If it didn't exit cleanly, can it be possible that the exception is the correct expected one, and that its message contains the search string? I'm probably missing the reason why this has to be checked here.

Comment (devaraj-kavali) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: With this PR change, SparkException will not be caught and thrown directly, and nothing writing to System.err and also no exit(exitedCleanly = false) in this case,  here we need to check the thrown exception message whether it has the expected searchString or not.

**Comment by srowen** [comment]
That's checked below. Why this change?

**Comment by devaraj-kavali** [comment]
This change is to avoid the expected exception is being thrown from the thread and getting printed in the test log.

**Comment by devaraj-kavali** [comment]
Yes, you are right, `exitedCleanly` is `false` even when the expected exception is thrown.

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala (hunk 1 above)
Comment (vanzin) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: I actually wonder if throwing this exception below is actually doing anything? `thread.join()` will not throw it in the test thread, which is what would actually matter.

So it seems that to achieve what the comment above says, you should do just this instead:

```
if (!exitedCleanly) {
  message = e.getMessage
}
```

Comment (devaraj-kavali) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: > I actually wonder if throwing this exception below is actually doing anything?

It would give the stack trace in the test log if there is any unexpected exception. With the below, we may give exception message but no stack trace in the log during fail.
```
if (!exitedCleanly) {
  message = e.getMessage
}
```

Comment (vanzin) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: Then just print the exception, or fail the test explicitly with that exception somehow. At least that's explicitly doing what is expected of it.

Comment (vanzin) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: (e.g., by stashing the exception in a variable and doing these checks on the test thread after the `thread.join()` call.)

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 2]
```diff
@@ -74,20 +74,27 @@ trait TestPrematureExit {
...
+        case e: Exception => exception = e
       }
     }
     thread.start()
     thread.join()
     val joined = printStream.lineBuffer.mkString("\n")
-    if (!joined.contains(searchString)) {
-      fail(s"Search string '$searchString' not found in $joined")
+    val searchStrContainsInEx = exception != null && exception.getMessage.contains(searchString)
+    if(!searchStrContainsInEx){
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: You need spaces around parens here. But I think there's a logic issue; if exception is `null` but `exitedCleanly` is `false`, you try to throw `null`. I get that it won't be null if so, but that highlights that I think they're redundant. And I don't think it's helping to pull out the condition here in a variable.

I think this is just simpler to delete `exitedCleanly` and ...

```
@volatile var: Exception = null
...
if (exception != null) {
  if (!exception.getMessage.contains(searchString)) {
    throw exception
  }
} else if (!joined.contains(searchString) {
  fail(...)
}
```
I think that's equivalent?

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 3]
```diff
@@ -74,20 +74,28 @@ trait TestPrematureExit {
...
+        case e: Exception => exception = e
       }
     }
     thread.start()
     thread.join()
     val joined = printStream.lineBuffer.mkString("\n")
-    if (!joined.contains(searchString)) {
-      fail(s"Search string '$searchString' not found in $joined")
+    val searchStrContainsInEx = exception != null && exception.getMessage.contains(searchString)
+    if (!searchStrContainsInEx) {
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: I thinks this is a bit hard to follow. Isn't this equivalent to:

```
if (exitedCleanly) {
  // check the output captured in printStream
} else {
  assert(exception != null)
  if (message does not contain search string) {
    throw exception
  }
}
```

Also, now that you moved the logic here, the comment in L83 is kinda orphaned, and a fresher comment explaining this block would be better.

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala (hunk 2 above)
Comment (devaraj-kavali) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: I've update as per this, thanks for the suggestion.

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala (hunk 3 above)
Comment (devaraj-kavali) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: I've updated as per Sean's comment, plz check whether it is ok.

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 4]
```diff
@@ -70,23 +70,25 @@ trait TestPrematureExit {
...
     }
     thread.start()
     thread.join()
     val joined = printStream.lineBuffer.mkString("\n")
-    if (!joined.contains(searchString)) {
+    if (exception != null) {
+      if (!exception.getMessage.contains(searchString)) {
+        throw exception
+      }
+    } else if (!joined.contains(searchString)) {
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: This could be just `assert(joined.contains(blah))`. You could also move the assignment of `joined` inside the `else` here.

Otherwise looks good pending tests.

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 5]
```diff
@@ -74,20 +74,25 @@ trait TestPrematureExit {
...
       }
     }
     thread.start()
     thread.join()
-    val joined = printStream.lineBuffer.mkString("\n")
-    if (!joined.contains(searchString)) {
-      fail(s"Search string '$searchString' not found in $joined")
+    if (exitedCleanly) {
+      val joined = printStream.lineBuffer.mkString("\n")
+      assert(joined.contains(searchString), s"Search string '$searchString' not found in $joined")
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: The message is redundant, `assert` already gives you a nice message, e.g.

```
[info]   "foo" did not contain "bar" (blah.scala)
```

**Review by vanzin — COMMENTED** [review]
Looks good pending tests.
